File: partiql-ast/src/main/java/org/partiql/ast/v1/AstNode.java
package org.partiql.ast.v1;

import org.jetbrains.annotations.NotNull;

import java.util.Collection;
import java.util.Random;

/**
 * TODO docs, equals, hashcode
 */
public abstract class AstNode {
    @NotNull
    public String tag = "Ast-" + String.format("%06x", new Random().nextInt());

    @NotNull
    public abstract Collection<AstNode> children();

    public abstract <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx);
}

File: partiql-ast/src/main/kotlin/org/partiql/ast/v1/AstNode.kt
package org.partiql.ast.v1

import kotlin.random.Random

/**
 * TODO docs, equals, hashcode
 */
public abstract class AstNode {
    @JvmField
    public var tag: String = "Ast-${"%06x".format(Random.nextInt())}"

    public abstract fun children(): Collection<AstNode>

    public abstract fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R
}
